presentation/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterBuilder.tsx

```diff
@@ -0,0 +1,157 @@
+  enableClassFilteringByProperties?: boolean;
+}
+
+/** @alpha */
+export function PresentationInstanceFilterBuilder(props: PresentationInstanceFilterBuilderProps) {
+  const {imodel, descriptor, onInstanceFilterChanged, enableClassFilteringByProperties} = props;
+  const classHierarchyProvider = useECClassHierarchyProvider(imodel);
+  const filteringProps = usePresentationInstanceFilteringProps(descriptor, classHierarchyProvider, enableClassFilteringByProperties);
+
+  const onFilterChanged = React.useCallback((filter?: PropertyFilter) => {
+    const presentationFilter = filter ? createPresentationInstanceFilter(descriptor, filter) : undefined;
+    onInstanceFilterChanged(presentationFilter);
+  }, [descriptor, onInstanceFilterChanged]);
+
+  return <InstanceFilterBuilder
+    {...filteringProps}
+    onFilterChanged={onFilterChanged}
+  />;
+}
+
+/** @alpha */
+export function usePresentationInstanceFilteringProps(descriptor: Descriptor, classHierarchyProvider?: ECClassHierarchyProvider, enableClassFiltering?: boolean) {
+  const [selectedClasses, setSelectedClasses] = React.useState<ClassInfo[]>([]);
+  const propertyInfos = React.useMemo(() => createInstanceFilterPropertyInfos(descriptor), [descriptor]);
+  const properties = React.useMemo(() => {
+    const matchingClassesSet = getClassesSet(selectedClasses.map((selectedClass) => selectedClass.id), classHierarchyProvider);
+    return propertyInfos
+      .filter((info) => !matchingClassesSet || info.sourceClassIds.some((id) => matchingClassesSet.has(id, {isDerived: true, isBase: true})))
+      .map((info) => info.propertyDescription);
+  }, [propertyInfos, selectedClasses, classHierarchyProvider]);
+
+  const classes = React.useMemo(() => descriptor.selectClasses.map((selectClass) => selectClass.selectClassInfo), [descriptor]);
+
+  React.useEffect(() => {
+    setSelectedClasses([]);
+  }, [descriptor]);
+
+  const onClassSelected = React.useCallback((classInfo: ClassInfo) => {
+    setSelectedClasses([...selectedClasses, classInfo]);
+  }, [selectedClasses]);
+
+  const onClassDeselected = React.useCallback((classInfo: ClassInfo) => {
+    setSelectedClasses(selectedClasses.filter((info) => info.id !== classInfo.id));
+  }, [selectedClasses]);
+
+  const onClearClasses = React.useCallback(() => {
+    setSelectedClasses([]);
+  }, []);
+
+  const onPropertySelected = React.useCallback((property: PropertyDescription) => {
+    if (!enableClassFiltering)
+      return;
+    const propertyInfo = propertyInfos.find((info) => info.propertyDescription.name === property.name);
+    if (!propertyInfo)
+      return;
+
+    const selectedClassesByProperty = computeSelectedClassesByProperty(propertyInfo, classes, selectedClasses, classHierarchyProvider);
+    if (selectedClassesByProperty)
+      setSelectedClasses(selectedClassesByProperty);
+  }, [classes, propertyInfos, selectedClasses, classHierarchyProvider, enableClassFiltering]);
```

Comment:
These callbacks produce new state that depends on the previous one. Please use the `setState` overload which supplies the previous state to the callback in order to avoid subtle bugs.